### ISSUE TITLE
Elassandra assertion error on bulk update

### ISSUE THREAD

**Issue by vibhash-singh** [strapdata/elassandra#271]
**Elassandra version**: 6.2.3.10

**Plugins installed**: []

**JVM version** (`java -version`):

```
 openjdk version "1.8.0_181"
OpenJDK Runtime Environment (build 1.8.0_181-8u181-b13-2~deb9u1-b13)
OpenJDK 64-Bit Server VM (build 25.181-b13, mixed mode)
```

**OS version** (`uname -a` if on a Unix-like system): Ubuntu 18.04

**Description of the problem including expected versus actual behavior**:
Conditional bulk update fails in elassandra but works in elasticsearch 6.2.3

**Steps to reproduce**:
1. Create an index and add data
`curl -XPOST -H 'content-type: application/json' http://172.17.0.3:9200/test_index/user -d '{"projects":["p1", "p2"]}'`

2. Run the below command 

```
curl -XPOST -H 'content-type: application/json'  http://172.17.0.3:9200/_bulk -d '
{"update":{"_index":"test_index","_type":"user","_id":"e8druWkBu-AlXJ3uggHp"}}
{"script":{"source": "if (!ctx._source.projects.contains(params.project)) {ctx._source.projects.add(params.project)}","params":{"project":"p3"}}}
'
'
```
3. Curl statement never gave any response back. I checked the result in elastic seach and found that data got updated but Cassandra system.log shows an AssertionError

```
2019-03-26 09:56:19,228 ERROR [elasticsearch[172.17.0.3][bulk][T#2]] CassandraDaemon.java:231 uncaughtException Exception in thread Thread[elasticsearch[172.17.0.3][bulk][T#2],5,main]
java.lang.AssertionError: failed result should not have a sequence number
        at org.elasticsearch.action.bulk.TransportShardBulkAction.processUpdateResponse(TransportShardBulkAction.java:289)
        at org.elasticsearch.action.bulk.TransportShardBulkAction.executeUpdateRequestOnce(TransportShardBulkAction.java:393)
        at org.elasticsearch.action.bulk.TransportShardBulkAction.executeUpdateRequest(TransportShardBulkAction.java:419)
        at org.elasticsearch.action.bulk.TransportShardBulkAction.executeBulkItemRequest(TransportShardBulkAction.java:250)
        at org.elasticsearch.action.bulk.TransportShardBulkAction.performOnPrimary(TransportShardBulkAction.java:129)
        at org.elasticsearch.action.bulk.TransportShardBulkAction.shardOperationOnPrimary(TransportShardBulkAction.java:115)
        at org.elasticsearch.action.bulk.TransportShardBulkAction.shardOperationOnPrimary(TransportShardBulkAction.java:73)
        at org.elasticsearch.action.support.replication.TransportReplicationAction$PrimaryShardReference.perform(TransportReplicationAction.java:1065)
        at org.elasticsearch.action.support.replication.TransportReplicationAction$PrimaryShardReference.perform(TransportReplicationAction.java:1043)
        at org.elasticsearch.action.support.replication.ReplicationOperation.execute(ReplicationOperation.java:104)
        at org.elasticsearch.action.support.replication.TransportReplicationAction$AsyncPrimaryAction.doRun(TransportReplicationAction.java:331)
        at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37)
        at org.elasticsearch.action.support.replication.TransportReplicationAction$PrimaryOperationTransportHandler.messageReceived(TransportReplicationAction.java:297)
        at org.elasticsearch.action.support.replication.TransportReplicationAction$PrimaryOperationTransportHandler.messageReceived(TransportReplicationAction.java:284)
        at org.elasticsearch.transport.RequestHandlerRegistry.processMessageReceived(RequestHandlerRegistry.java:66)
        at org.elasticsearch.transport.TransportService$7.doRun(TransportService.java:661)
        at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:672)
        at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)

```

**Please provide the following information:

* elassandra logs (logs/system.logs or /var/lib/cassandra/system.log) : Added above
* elasticsearch cluster state (curl http://localhost:9200/_cluster/state): 
```
{
  "cluster_name": "Test Cluster",
  "compressed_size_in_bytes": 501,
  "version": 9,
  "state_uuid": "QtmmOC9dS8SMAGeppPEduw",
  "master_node": "a8feb81e-270f-4def-8216-4f179bd1f5e2",
  "blocks": {},
  "nodes": {
    "a8feb81e-270f-4def-8216-4f179bd1f5e2": {
      "name": "172.17.0.3",
      "status": "ALIVE",
      "ephemeral_id": "a8feb81e-270f-4def-8216-4f179bd1f5e2",
      "transport_address": "172.17.0.3:9300",
      "attributes": {
        "rack": "r1",
        "dc": "DC1"
      }
    }
  },
  "metadata": {
    "version": 2,
    "cluster_uuid": "a8feb81e-270f-4def-8216-4f179bd1f5e2",
    "templates": {},
    "indices": {
      "test_index": {
        "state": "open",
        "settings": {
          "index": {
            "creation_date": "1553594023962",
            "number_of_shards": "1",
            "number_of_replicas": "0",
            "uuid": "jKfz0fVyRnKTBrsNuBZeQQ",
            "version": {
              "created": "6020399"
            },
            "provided_name": "test_index"
          }
        },
        "mappings": {
          "user": {
            "properties": {
              "projects": {
                "type": "text",
                "fields": {
                  "keyword": {
                    "ignore_above": 256,
                    "type": "keyword"
                  }
                }
              }
            }
          }
        },
        "aliases": [],
        "primary_terms": {
          "0": 0
        },
        "in_sync_allocations": {
          "0": []
        }
       }
    },
    "index-graveyard": {
      "tombstones": []
    }
  },
  "routing_table": {
    "indices": {
      "test_index": {
        "shards": {
          "0": [
            {
              "state": "STARTED",
              "primary": true,
              "node": "a8feb81e-270f-4def-8216-4f179bd1f5e2",
              "relocating_node": null,
              "shard": 0,
              "index": "test_index",
              "token_ranges": [
                "(-9223372036854775808,9223372036854775807]"
              ],
              "allocation_id": {
                "id": "dummy_alloc_id"
              }
            }
          ]
        }
      }
    }
  },
  "routing_nodes": {
    "unassigned": [],
    "nodes": {
      "a8feb81e-270f-4def-8216-4f179bd1f5e2": [
        {
          "state": "STARTED",
          "primary": true,
          "node": "a8feb81e-270f-4def-8216-4f179bd1f5e2",
          "relocating_node": null,
          "shard": 0,
          "index": "test_index",
          "token_ranges": [
            "(-9223372036854775808,9223372036854775807]"
          ],
          "allocation_id": {
            "id": "dummy_alloc_id"
          }
        }
      ]
    }
  },
  "snapshot_deletions": {
    "snapshot_deletions": []
  },
  "snapshots": {
    "snapshots": []
  },
  "restore": {
    "snapshots": []
  }
}

```

* cassandra schema (cqlsh>DESC KEYSPACE <your_keyspace>)
```
CREATE KEYSPACE test_index WITH replication = {'class': 'NetworkTopologyStrategy', 'DC1': '1'}  AND durable_writes = true;
CREATE TABLE test_index.user (
    "_id" text PRIMARY KEY,
    projects list<text>
) WITH bloom_filter_fp_chance = 0.01
    AND caching = {'keys': 'ALL', 'rows_per_partition': 'NONE'}
    AND comment = ''
    AND compaction = {'class': 'org.apache.cassandra.db.compaction.SizeTieredCompactionStrategy', 'max_threshold': '32', 'min_threshold': '4'}
    AND compression = {'chunk_length_in_kb': '64', 'class': 'org.apache.cassandra.io.compress.LZ4Compressor'}
    AND crc_check_chance = 1.0
    AND dclocal_read_repair_chance = 0.1
    AND default_time_to_live = 0
    AND gc_grace_seconds = 864000
    AND max_index_interval = 2048
    AND memtable_flush_period_in_ms = 0
    AND min_index_interval = 128
    AND read_repair_chance = 0.0
    AND speculative_retry = '99PERCENTILE';
CREATE CUSTOM INDEX elastic_user_idx ON test_index.user () USING 'org.elassandra.index.ExtendedElasticSecondaryIndex';
```

* cassandra gossip state (run: nodetool gossipinfo)

```
/172.17.0.3
  generation:1553592898
  heartbeat:1706
  STATUS:55:NORMAL,-2180626626081522229
  LOAD:1661:113785.0
  SCHEMA:1208:ab8b61c6-d700-3c2e-ae37-51b6bc17d07b
  DC:8:DC1
  RACK:10:r1
  RELEASE_VERSION:4:3.11.3.5
  INTERNAL_IP:6:172.17.0.3
  RPC_ADDRESS:3:172.17.0.3
  NET_VERSION:1:11
  HOST_ID:2:a8feb81e-270f-4def-8216-4f179bd1f5e2
  RPC_READY:67:true
  X1:1205:{"test_index":3}
  X2:1210:a8feb81e-270f-4def-8216-4f179bd1f5e2/2
  TOKENS:54:<hidden>
```

**Comment by vroyer** [comment]
Hi,

Yes, there is an issue with such update.
You can also do the same thing by mapping your field projects to a cassandra set,  and add p3 to project with a CQL statement : UPDATE users SET projects = projects + 'p3' WHERE ...

Moreover, keep in mind that update script on an eventually consistent store like Cassandra could give inconsistent results.

**Comment by vroyer** [comment]
For more information, see cql_collection settings in http://doc.elassandra.io/en/latest/mapping.html#type-mapping

**Comment by vroyer** [comment]
This issue was fix in 6.2.3.13.
